protocols/gossipsub/src/handler.rs

```diff
@@ -407,10 +411,19 @@ impl EnabledHandler {
             }
         }
 
-        // Drop the next message in queue if it's stale.
-        if let Poll::Ready(Some(rpc)) = self.send_queue.poll_stale(cx) {
+        // Remove stale messages from the queue.
+        let stale = self.message_queue.retain_mut(|rpc| match rpc {
+            RpcOut::Publish {
+                ref mut timeout, ..
+            }
+            | RpcOut::Forward {
+                ref mut timeout, ..
+            } => !timeout.poll_unpin(cx).is_ready(),
+            _ => true,
+        });
```

Comment:
I know that this is existing logic, but do we really need to remove stale messages from the queue here?
It results the the whole internal binary heap of the queue being rebuild every single time the handler is polled. Isn't it enough that the handler simply checks the timeout before sending out stale messages?

I know it shrinks the queue length, but I wonder how often in the real world a message actually times out, and whether in that case there is a larger issue at hand anyway. Do you have any data on that?

Reply:
This is a good point. 

In terms of data, we often see non-priority messages being timed out on slow machines, or if bandwidth isn't adequate. But to your point, we can probably drop them when we pop them, without any harm.

Reply:
@jxs - I think potentially the performance cost of this retain isn't worth it. Maybe should just drop the message when we read from the queue.

Reply:
I guess you wanted to group all the stale messages. I feel like a O(n) search every poll might not be worth the grouping tho? What do you think?